gnd/src/main/java/com/google/android/gnd/rx/Result.java

```diff
@@ -0,0 +1,164 @@
+   * </pre>
+   *
+   * @param fn A function that takes an argument of type T and returns a Single with emissions of
+   *     type R.
+   * @param <T> The initial type of values passed to fn.
+   * @param <R> The type of emissions produced by the Single returned by fn.
+   * @return
+   */
+  public static <T, R> Function<T, Single<Result<R>>> mapSingle(
+      @NonNull Function<T, Single<R>> fn) {
+    return (T value) -> fn.apply(value).map(Result::success).onErrorReturn(Result::error);
+  }
+
+  /**
+   * Wraps single emissions in a Result without applying any other transformations. Use this
+   * function to wrap and handle potential errors at a given point in a chain of RX transformations.
+   *
+   * @param single
+   * @param <T> The type of the Single's emissions.
+   * @return A Single that emits Results.
+   */
+  public static <T> Single<Result<T>> wrapSingle(Single<T> single) {
+    return single.map(Result::success).onErrorReturn(Result::error);
+  }
+
+  /**
+   * Transforms an Observable, returned by applying fn to an argument of type T, into an Observable
+   * that emits Results, returning Result::error on error.
+   *
+   *<p>This function is useful for handling nested RX streams, for example, in situations in which
+   * a function passed to the RX {@code map} operator returns a Observable. {@code mapObservable} ensures we
+   * handle errors produced by such 'interior' Observables without breaking the 'exterior' stream.
+   *
+   * <pre>
+   *     public Observable<Project> getProject(String projectId) {
+   *        return dataRepository.getProject(projectId);
+   *     }
+   *
+   *     somePublishSubject.onNext("foobar");
+   *
+   *     somePublishSubject.map(Result.mapObservable(getProject))
+   *                       .subscribe(Result.unwrap(onProjectSuccess, onProjectError))
+   * </pre>
+   *
+   *
+   * @param fn A function that takes an argument of type T and returns an Observable with emissions
+   *     of type R.
+   * @param <T> The initial type of values passed to fn.
+   * @param <R> The type of emissions produced by the Observable returned by fn.
+   * @return
+   */
+  public static <T, R> Function<T, Observable<Result<R>>> mapObservable(
+      @NonNull Function<T, Observable<R>> fn) {
+    return (T value) -> fn.apply(value).map(Result::success).onErrorReturn(Result::error);
+  }
+
+  /**
+   * Wraps observable emissions in a Result without applying any other transformations. Use this
+   * function to wrap and handle potential errors at a given point in a chain of RX transformations.
+   *
+   * @param observable
+   * @param <T> The type of the observable's emissions.
+   * @return An Observable that emits Results.
+   */
+  public static <T> Observable<Result<T>> wrapObservable(Observable<T> observable) {
+    return observable.map(Result::success).onErrorReturn(Result::error);
+  }
+
+  public static <T> Consumer<? super Result<T>> unwrap(
```

Comment:
Missing Javadoc here, but I'd like to unblock this PR so we can add it in a follow-up.

Reply:
Doh! will do.

Reply:
Going forward I'd add only essential Javadoc to public methods; inferred types probably don't need to be described since the caller doesn't explicitly declare them, and when params and return values are already described by the main text we can omit them as well.